src/arithmetic/arithmetic_expression_construct.c

```diff
@@ -441,6 +443,96 @@ static AR_ExpNode *_AR_ExpFromNamedPath(const cypher_astnode_t *path) {
+	uint path_len = cypher_ast_pattern_path_nelements(path);
+	if(path_len != 3) {
+		ErrorCtx_SetError("shortestPath requires a path containing a single relationship");
+		return AR_EXP_NewConstOperandNode(SI_NullVal());
+	}
+
+	// Retrieve the minimum and maximum number of hops, if specified.
+	int start = 1;
+	int end = EDGE_LENGTH_INF;
+	const cypher_astnode_t *edge = cypher_ast_pattern_path_get_element(path, 1);
+	const cypher_astnode_t *range = cypher_ast_rel_pattern_get_varlength(edge);
+	if(range == NULL) {
+		end = 1; // Not a variable-length edge
+	} else {
+		const cypher_astnode_t *range_start = cypher_ast_range_get_start(range);
+		if(range_start) {
+			// If specified, the edge's minimum hop value must be 0 or 1
+			start = AST_ParseIntegerNode(range_start);
+			if(start > 1) {
+				ErrorCtx_SetError("shortestPath does not support a minimal length different from 0 or 1");
+				return AR_EXP_NewConstOperandNode(SI_NullVal());
+			}
+		}
+		const cypher_astnode_t *range_end = cypher_ast_range_get_end(range);
+		if(range_end) end = AST_ParseIntegerNode(range_end);
+	}
+
+	if(!cypher_ast_shortest_path_is_single(path)) {
+		ErrorCtx_SetError("RedisGraph does not currently support allShortestPaths");
+		return AR_EXP_NewConstOperandNode(SI_NullVal());
+	}
+
+	enum cypher_rel_direction dir = cypher_ast_rel_pattern_get_direction(edge);
+	if(dir == CYPHER_REL_BIDIRECTIONAL) {
+		ErrorCtx_SetError("RedisGraph does not currently support undirected shortestPath traversals");
+		return AR_EXP_NewConstOperandNode(SI_NullVal());
+	}
+
+	if(cypher_ast_rel_pattern_get_properties(edge)) {
+		ErrorCtx_SetError("RedisGraph does not currently support filters on relationships in shortestPath");
+		return AR_EXP_NewConstOperandNode(SI_NullVal());
+	}
+
+	// Collect the IDs of all relationship types
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	uint reltype_count = cypher_ast_rel_pattern_nreltypes(edge);
+	int *reltypes = NULL;
+	if(reltype_count > 0) {
+		reltypes = array_new(int, reltype_count);
+		for(uint i = 0; i < reltype_count; i ++) {
+			const char *reltype = cypher_ast_reltype_get_name(cypher_ast_rel_pattern_get_reltype(edge, i));
+			Schema *s = GraphContext_GetSchema(gc, reltype, SCHEMA_EDGE);
+			if(!s) {
+				// Encountered unknown relationship
+				ErrorCtx_SetError("Encountered unknown relationship type '%s' in shortestPath", reltype);
+				array_free(reltypes);
+				return AR_EXP_NewConstOperandNode(SI_NullVal());
+			}
+			reltypes = array_append(reltypes, s->id);
+		}
+	}
+
+	AR_ExpNode *op = AR_EXP_NewOpNode("shortestpath", 2);
+
+	// Instantiate a context struct with traversal details.
+	ShortestPathCtx *ctx = rm_malloc(sizeof(ShortestPathCtx));
+	ctx->minHops = start;
+	ctx->maxHops = end;
+	ctx->reltypes = reltypes;
+	ctx->R = GrB_NULL;
+	ctx->TR = GrB_NULL;
+	ctx->free_matrices = false;
+
+	// Add the context to the function descriptor as the function's private data.
+	op->op.f = AR_SetPrivateData(op->op.f, ctx);
```

Comment:
I think we need to find a better way to associate a function private data / context object, maybe passing it to `AR_EXP_NewOpNode` ? or introducing a new function e.g. `AR_EXP_NewOpWithCtxNode` ??

Reply:
I agree, this could be improved on. I don't think I have an opinion on whether it would be better to extend `AR_EXP_NewOpNode` or introduce a new function.